doc/decisions/1_in_discussion/reference_counting.md

```diff
@@ -0,0 +1,286 @@
+In an ARC system, every heap memory allocation has a reference count, which is often stored as a hidden header.
+Essentially, instead of calling `malloc(data_size)` the system does `malloc(header_size+data_size)` and uses the additional `header_size` bytes to store data like the reference count.
+Crucially, the pointer returned to the caller still points after the header data, which is therefore hidden.
+
+Since (almost) all allocations for Elektra's data structures go through `elektraMalloc`, we could implement something similar.
+In `elektraMalloc` we allocate a few extra bytes (for alignment reasons this would be `sizeof(size_t)`) and use them for reference counting.
+The reference counting would be done via generic `elektraRetain` and `elektraRelease` functions.
+These functions increment and decrement the counter in the hidden header.
+Additionally, `elektraRelease` would also call `free` when the count reaches zero.
+
+Like in the "Retain/Release/Free" option above, the hidden header data could also be used to store a pointer to a `free` function.
+Then things like `mmapstorage` could be covered as well.
+
+```c
+// Based on: https://codereview.stackexchange.com/questions/146561/reference-counting-in-c99
+
+// Note: should NOT call free() on pointer
+typedef void (*Destructor)(const void*);
+
+typedef struct {
+  Destructor destroy;
+  unsigned int count;
+} ObjectRecord;
+
+static ObjectRecord *toRecord(void *object) { return (ObjectRecord*)(object) - 1; }
+static void *fromRecord(ObjectRecord *record) { return record + 1; }
+
+void * elektraMalloc(size_t size, ) {
+  ObjectRecord *record = malloc (sizeof (ObjectRecord) + size);
+  if (record == NULL) {
+    return NULL;
+  }
+  record->count = 1;
+  record->destroy = destroy;
+  return recordToObject (record);
+}
+
+unsigned int elektraRefCount(const void * p) {
+  return objectToRecord (p)->count;
+}
+
+void elektraRetain(const void *p)
+{
+    ++objectToRecord(p)->count;
+}
+
+void elektraRelease(const void *p)
+{
+    ObjectRecord *record = objectToRecord(p);
+
+    record->count--;
+    if (record->count > 0) {
+        /* still referenced */
+        return;
+    }
+
+    if (record->destory) {
+        record->destory(p);
+    }
+    free(record);
+}
+```
+
+Usage example:
+
+```c
+static void ksDestroy(const void * p) {
+  KeySet * ks = (KeySet *)p;
+  for (size_t i = 0; i < ks->size; i++) {
+    elektraRelease(key);
+  }
+  // [...]
+}
+
+KeySet * ksCreate() {
+  KeySet * ks = elektraMalloc(sizeof(struct KeySet), ksDestroy);
+  // [...] initalization
+  return ks;
+}
+
+void ksAppendKey(KeySet * ks, Key * key) {
+  elektraRetain(key);
+  // [...]
+}
+
+// for compatibility and for complete API
+static inline void ksDel(KeySet * ks) {
+  elektraRelease(ks);
+}
+```
+
+## Decision
+
+General decisions regardless of implementation:
+
+- `Key` has two counters.
+  One that blocks deletion, another to prevent name changes.
```

Comment:
No it hasn't yet. But it was already decided that we'll use a second counter with this exact "reentrant lock" idea to automatically lock/unlock the key when it is/isn't in a KeySet. In that case, I don't think there are many alternatives, since it is the prime example use case for a reentrant lock.

Reply:
We talked talked past each other when we discussed this. I never thought of the second ref counter as good idea but argued that we should try to simplify the problem (e.g. stay read-only, stay locked etc.)

Reply:
Oh okay, you should've been more clear on that.... The old decision was already in "Partially implemented" and had "add second counter to Key" in the "Decision" section.

I don't think this is the right place to discuss it. Since this PR is not at all about the locking of keynames. (I will remove this note about two counters). The keyname locking stuff should be discussed in the next PR for https://github.com/ElektraInitiative/libelektra/blob/master/doc/decisions/0_drafts/readonly_keynames.md